database/database_internal_test.go

```diff
@@ -415,6 +416,170 @@ func TestLayers(t *testing.T) {
+		}, AosVersion: uint64(i), UnitSubject: true, UID: i + 100}
+
+		subjectInstance.InstanceID = fmt.Sprintf("%s_%s_%s", subjectInstance.ServiceID, subjectInstance.SubjectID,
+			strconv.Itoa(i))
+
+		if i%2 == 0 {
+			subjectInstance.Running = true
+			running = append(running, subjectInstance)
+		}
+
+		if err := db.AddInstance(subjectInstance); err != nil {
+			t.Fatalf("Can't add instance to DB %v", err)
+		}
+
+		subjectInstances = append(subjectInstances, subjectInstance)
+		allInstances = append(allInstances, subjectInstance)
+
+		serviceInstance := launcher.InstanceInfo{InstanceIdent: cloudprotocol.InstanceIdent{
+			ServiceID: testServiceID,
+			SubjectID: "someSubject" + strconv.Itoa(i), Instance: uint64(i),
+		}, AosVersion: uint64(i + 20), UnitSubject: true, UID: i + 200}
+
+		serviceInstance.InstanceID = fmt.Sprintf("%s_%s_%s", serviceInstance.ServiceID, serviceInstance.SubjectID,
+			strconv.Itoa(i))
+
+		if i%2 == 0 {
+			serviceInstance.Running = true
+			running = append(running, serviceInstance)
+		}
+
+		if err := db.AddInstance(serviceInstance); err != nil {
+			t.Fatalf("Can't add instance to DB %v", err)
+		}
+
+		serviceInstances = append(serviceInstances, serviceInstance)
+		allInstances = append(allInstances, serviceInstance)
+	}
+
+	allResults, err := db.GetAllInstances()
+	if err != nil {
+		t.Fatalf("Can't get all instances from DB %v", err)
+	}
+
+	if !reflect.DeepEqual(allResults, allInstances) {
+		t.Error("Incorrect get all instances result")
+	}
+
+	runningResult, err := db.GetRunningInstances()
+	if err != nil {
+		t.Fatalf("Can't get running instances from DB %v", err)
+	}
+
+	if !reflect.DeepEqual(runningResult, running) {
+		t.Error("Incorrect running instances result")
+	}
+
+	serviceInstancesResult, err := db.GetServiceInstances(testServiceID)
+	if err != nil {
+		t.Fatalf("Can't get instances by serviceID %v", err)
+	}
+
+	if !reflect.DeepEqual(serviceInstancesResult, serviceInstances) {
+		t.Error("Incorrect instances by serviceID result")
+	}
+
+	serviceInstancesResult, err = db.GetServiceInstances("notAvailableServiceID")
+	if err != nil {
+		t.Fatalf("Can't get instances by serviceID %v", err)
+	}
+
+	if len(serviceInstancesResult) != 0 {
+		t.Error("incorrect count of instances")
+	}
+
+	subjectInstancesResult, err := db.GetSubjectInstances(testSubjectID)
+	if err != nil {
+		t.Fatalf("Can't get instances by subjectID %v", err)
+	}
+
+	if !reflect.DeepEqual(subjectInstancesResult, subjectInstances) {
+		t.Error("Incorrect instances by subjectID result")
+	}
+
+	testInstanceInfo := launcher.InstanceInfo{InstanceIdent: cloudprotocol.InstanceIdent{
+		ServiceID: testServiceID,
+		SubjectID: testSubjectID, Instance: 42,
+	}, AosVersion: 42, UnitSubject: true, UID: 42}
+
+	testInstanceID := fmt.Sprintf("%s_%s_%s", testInstanceInfo.ServiceID, testInstanceInfo.SubjectID, strconv.Itoa(42))
+
+	testInstanceInfo.InstanceID = testInstanceID
+
+	if err := db.AddInstance(testInstanceInfo); err != nil {
```

Comment:
Move negative tests into separate test or add comments what is tested.

Reply:
added comment